src/coprocessor/codec/mysql/duration.rs

```diff
@@ -368,127 +478,208 @@ impl Duration {
+                let mut hours = self.hours() + rhs.hours();
+
+                if i64::from(micros) >= MICROS_PER_SEC {
+                    micros -= MICROS_PER_SEC as u32;
+                    secs += 1;
+                }
+                if secs >= 60 {
+                    secs -= 60;
+                    minutes += 1;
+                }
+                if minutes >= 60 {
+                    minutes -= 60;
+                    hours += 1;
+                }
+
+                check_hour(hours).ok()?;
+
+                Some(Duration::new(
+                    false,
+                    hours,
+                    minutes,
+                    secs,
+                    micros,
+                    self.fsp().max(rhs.fsp()),
+                ))
+            }
+        }
     }
 
     /// Checked duration subtraction. Computes self - rhs, returning None if overflow occurred.
-    pub fn checked_sub(self, rhs: &Duration) -> Option<Duration> {
-        let sub = match self.to_nanos().checked_sub(rhs.to_nanos()) {
-            Some(result) => result,
-            None => return None,
-        };
-        Duration::from_nanos(sub, self.fsp().max(rhs.fsp()) as i8).ok()
+    pub fn checked_sub(self, rhs: Duration) -> Option<Duration> {
+        match (self.get_neg(), rhs.get_neg()) {
+            (false, true) => self.checked_add(rhs.abs()),
+            (true, false) => self.abs().checked_add(rhs.abs()).map(|mut res| {
+                res.set_neg(true);
+                res
+            }),
+            (true, true) => rhs.abs().checked_sub(self.abs()),
+            (false, false) => {
+                let neg = self < rhs;
+
+                let (l, r) = if neg { (rhs, self) } else { (self, rhs) };
+
+                let mut micros = l.micros() as i32 - r.micros() as i32;
+                let mut secs = l.secs() as i32 - r.secs() as i32;
+                let mut minutes = l.minutes() as i32 - r.minutes() as i32;
+                let mut hours = l.hours() as i32 - r.hours() as i32;
+
+                if micros < 0 {
+                    micros += MICROS_PER_SEC as i32;
+                    secs -= 1;
+                }
+
+                if secs < 0 {
+                    secs += 60;
+                    minutes -= 1;
+                }
+
+                if minutes < 0 {
+                    minutes += 60;
+                    hours -= 1;
+                }
+
+                Some(Duration::new(
+                    neg,
+                    hours as u32,
+                    minutes as u32,
+                    secs as u32,
+                    micros as u32,
+                    self.fsp().max(rhs.fsp()),
+                ))
+            }
+        }
     }
-}
 
-impl Display for Duration {
-    fn fmt(&self, formatter: &mut Formatter<'_>) -> fmt::Result {
-        if self.neg {
-            write!(formatter, "-")?;
+    fn format(self, sep: &str) -> String {
+        let mut string = String::new();
```

Comment:
How about use `String::with_capacity`, and check `fsp` before allocate the string

Reply:
Use `write!` might be better.